File: src/main/clojure/futures.clj

;; Demo 12: Clojure Futures

(def f (future
         (Thread/sleep 3000)
         (.getName (Thread/currentThread))
         (+ 10 1000)
         ))

(println (realized? f))

(Thread/sleep 4000)

(println (realized? f))

(println (deref f))


;; Demo 13: Clojure Promise

(def p (promise))

(deliver p 10)

@p

;(deref p)

;(shutdown-agents)

File: src/main/scala/futures.scala

/**
  * Demo 10: Scala Futures
  */
import scala.concurrent._
import ExecutionContext.Implicits.global

val future = Future { x + 1 }

future.foreach(x -> println(x))

/**
  * Demo 11: Scala Promises
  */
val promise = Promise[Int]()
val futureFromPromise = promise.future

futureFromPromise.foreach(println)

promise.success(100)
